packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.test.tsx

```diff
@@ -0,0 +1,633 @@
+        fireEvent.click(screen.getByTestId('submit'));
+      });
+      expect(apiClientOverrides.apiCreateCustomer).toHaveBeenCalled();
+      expect(
+        screen.queryByTestId('error-payment-submission')
+      ).toBeInTheDocument();
+      expect(refreshSubscriptions).toHaveBeenCalledTimes(0);
+    });
+
+    it('displays for unexpected payment intent status', async () => {
+      const subscriptionResult = deepCopy(SUBSCRIPTION_RESULT);
+      subscriptionResult.latest_invoice.payment_intent.status =
+        'banana_deck_chair';
+      const apiClientOverrides = {
+        ...defaultApiClientOverrides(),
+        apiCreateSubscriptionWithPaymentMethod: jest
+          .fn()
+          .mockResolvedValue(subscriptionResult),
+      };
+      const { refreshSubscriptions } = await commonSubmitSetup({
+        apiClientOverrides,
+      });
+      await act(async () => {
+        fireEvent.click(screen.getByTestId('submit'));
+      });
+      await waitForExpect(() =>
+        expect(
+          apiClientOverrides.apiCreateSubscriptionWithPaymentMethod
+        ).toHaveBeenCalled()
+      );
+      expect(
+        screen.queryByTestId('error-payment-submission')
+      ).toBeInTheDocument();
+      expect(refreshSubscriptions).toHaveBeenCalledTimes(0);
+    });
+
+    it('displays if client secret is missing from payment intent', async () => {
+      const subscriptionResult = deepCopy(SUBSCRIPTION_RESULT);
+      subscriptionResult.latest_invoice.payment_intent.status =
+        'requires_action';
+      subscriptionResult.latest_invoice.payment_intent.client_secret = null;
+
+      const apiClientOverrides = {
+        ...defaultApiClientOverrides(),
+        apiCreateSubscriptionWithPaymentMethod: jest
+          .fn()
+          .mockResolvedValue(subscriptionResult),
+      };
+      const { stripeOverride, refreshSubscriptions } = await commonSubmitSetup({
+        apiClientOverrides,
+      });
+
+      await act(async () => {
+        fireEvent.click(screen.getByTestId('submit'));
+      });
+
+      await waitForExpect(() =>
+        expect(
+          apiClientOverrides.apiCreateSubscriptionWithPaymentMethod.mock
+            .calls[0][0]
+        ).toMatchObject({
+          // idempotencyKey (ignored)
+          priceId: PLAN.plan_id,
+          paymentMethodId: PAYMENT_METHOD_RESULT.paymentMethod.id,
+        })
+      );
+      await waitForExpect(() =>
+        expect(
+          screen.queryByTestId('error-payment-submission')
+        ).toBeInTheDocument()
+      );
+      expect(refreshSubscriptions).toHaveBeenCalledTimes(0);
+      expect(stripeOverride.confirmCardPayment).not.toHaveBeenCalled();
+    });
+  });
+
+  describe('errors', () => {
+    // We map Stripe error codes to less-specific error message string IDs for display.
+    // L10n is disabled for these error messages in testing, so they show up as string IDs.
+    // The codes are mapped to IDs in lib/errors.ts and the strings are in public/locales
```

Comment:
These comments are very helpful.  Perhaps we should move/copy them to where errors mapped/handled.

Reply:
Well, my intent for the comment here is to explain why we're asserting for codes like `getByText('card-error')` rather than `getByText('Your transaction could not be processed. Please verify your credit card information and try again.
')` - because l10n is disabled here and that skips a couple levels of indirection